kv/kvserver/kvflowcontrol/rac2: TestRangeController failed

kv/kvserver/kvflowcontrol/rac2.TestRangeController [failed](https://tanzanite.cluster.engflow.com/invocations/default/dd4879a5-ea10-4f8d-9619-20bbf1925c1e?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIva3ZmbG93Y29udHJvbC9yYWMyOnJhYzJfdGVzdA==) on master @ [dcce4cafa234525fc859d32745c11ed87890dc7b](https://github.com/cockroachdb/cockroach/commits/dcce4cafa234525fc859d32745c11ed87890dc7b):


```
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1284 rac2.TestRangeController.func1.1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 datadriven.RunTest.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 datadriven.runDirective.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:341 datadriven.runDirective ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:212 datadriven.runDirectiveOrSubTest ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:182 datadriven.runTestInternal ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 datadriven.RunTestAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:119 datadriven.RunTest ???
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1333 rac2.TestRangeController.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 datadriven.Walk.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 datadriven.WalkAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 datadriven.WalkAny.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:26 datadriven.subTest.func1 ???

goroutine 1319 lock 0xc0002fa248
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:209 rac2.(*testingRCState).sendStreamString ??? <<<<<
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:208 rac2.(*testingRCState).sendStreamString ???
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1284 rac2.TestRangeController.func1.1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 datadriven.RunTest.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 datadriven.runDirective.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:341 datadriven.runDirective ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:212 datadriven.runDirectiveOrSubTest ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:182 datadriven.runTestInternal ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 datadriven.RunTestAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:119 datadriven.RunTest ???
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1333 rac2.TestRangeController.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 datadriven.Walk.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 datadriven.WalkAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 datadriven.WalkAny.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:26 datadriven.subTest.func1 ???

goroutine 1319 lock 0xc0002fa368
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:209 rac2.(*testingRCState).sendStreamString ??? <<<<<
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:208 rac2.(*testingRCState).sendStreamString ???
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1284 rac2.TestRangeController.func1.1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 datadriven.RunTest.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 datadriven.runDirective.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:341 datadriven.runDirective ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:212 datadriven.runDirectiveOrSubTest ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:182 datadriven.runTestInternal ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 datadriven.RunTestAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:119 datadriven.RunTest ???
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1333 rac2.TestRangeController.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 datadriven.Walk.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 datadriven.WalkAny ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 datadriven.WalkAny.func1 ???
github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:26 datadriven.subTest.func1 ???



```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeController.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-42798

kv/kvserver/kvflowcontrol/rac2.TestRangeController [failed](https://tanzanite.cluster.engflow.com/invocations/default/7030b7e0-7daf-44c6-85df-47e72329a78e?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXIva3ZmbG93Y29udHJvbC9yYWMyOnJhYzJfdGVzdA==) on master @ [58c475d67e32b75284b4fe293bff82807c3d129d](https://github.com/cockroachdb/cockroach/commits/58c475d67e32b75284b4fe293bff82807c3d129d):

Fatal error:

```
panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked [recovered]
	panic: mutex is not write locked

```
Stack: 

```
goroutine 85 [running]:
testing.tRunner.func1.2({0x26b59a0, 0x2ef9bc0})
	GOROOT/src/testing/testing.go:1631 +0x3f7
testing.tRunner.func1()
	GOROOT/src/testing/testing.go:1634 +0x6b6
panic({0x26b59a0?, 0x2ef9bc0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc000a065d0?, {0x2f1ae58, 0x5208f40})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:226 +0x74
panic({0x26b59a0?, 0x2ef9bc0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0002383c0, {0x2f1ae58, 0x5208f40})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:535 +0x28b
panic({0x26b59a0?, 0x2ef9bc0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/datadriven.runDirective.func1.1()
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:334 +0x154
panic({0x26b59a0?, 0x2ef9bc0?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/util/syncutil.(*RWMutex).AssertHeld(...)
	github.com/cockroachdb/cockroach/pkg/util/syncutil/mutex_sync_race.go:131
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.(*rangeController).updateSendQueueStatsRaftMuRCLocked(0xc000434420, {0xc000bb8e70?, 0xc000a07358?, 0x0?})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller.go:1433 +0x4e8
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.TestRangeController.func1.1(0xc0002201a0, 0xc00042e898)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:1325 +0x450d
github.com/cockroachdb/datadriven.RunTest.func1({0x2f34d30, 0xc0002201a0}, 0xc00042e898)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:118 +0x59
github.com/cockroachdb/datadriven.runDirective.func1({0x2f34d30, 0xc0002201a0}, 0xc00042e898, 0xc00044c290)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:337 +0x10d
github.com/cockroachdb/datadriven.runDirective({0x2f34d30, 0xc0002201a0}, 0xc00042e870, 0xc00044c290)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:342 +0x90
github.com/cockroachdb/datadriven.runDirectiveOrSubTest({0x2f34d30, 0xc0002201a0}, 0xc00042e870, {0x0, 0x0}, 0xc00044c290)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:211 +0xa5
github.com/cockroachdb/datadriven.runTestInternal({0x2f34d30, 0xc0002201a0}, {0xc00077b0e0, 0xf0}, {0x2eff2a0, 0xc000078298}, 0xc00044c290, 0x0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:184 +0xe5
github.com/cockroachdb/datadriven.RunTestAny({0x2f34d30, 0xc0002201a0}, {0xc00077b0e0, 0xf0}, 0xc00044c290)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:144 +0x3d1
github.com/cockroachdb/datadriven.RunTest(0xc0002201a0, {0xc00077b0e0, 0xf0}, 0xc000bb8f00)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:117 +0xbb
github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2.TestRangeController.func1(0xc0002201a0, {0xc00077b0e0, 0xf0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go:962 +0x285
github.com/cockroachdb/datadriven.Walk.func1({0x2f34d30, 0xc0002201a0}, {0xc00077b0e0, 0xf0})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:425 +0x5f
github.com/cockroachdb/datadriven.WalkAny({0x2f34d30, 0xc0002201a0}, {0xc00077b0e0, 0xf0}, 0xc00031ed70)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:436 +0x282
github.com/cockroachdb/datadriven.WalkAny.func1({0x2f34d30, 0xc0002201a0})
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/datadriven.go:449 +0xf9
github.com/cockroachdb/datadriven.subTest.func1(0xc0002201a0)
	github.com/cockroachdb/datadriven/external/com_github_cockroachdb_datadriven/test_interface.go:27 +0x3c
testing.tRunner(0xc0002201a0, 0xc00031fd50)
	GOROOT/src/testing/testing.go:1689 +0x21f
created by testing.(*T).Run in goroutine 77
	GOROOT/src/testing/testing.go:1742 +0x826
```
<details><summary>Log preceding fatal error</summary>
<p>

```
        eval original in send-q: reg=+1.0 MiB ela=+0 B
        ++++
        MsgApps sent in pull mode:
         to: 2, lowPri: false entries: [1]
         to: 3, lowPri: false entries: [1]
        ++++
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/rac2/rac2_test_/rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/force_flush_choice:93:
        set_replicas [1 args]
        range_id=1 tenant_id=1 local_replica_id=1 next_raft_index=2
          store_id=1 replica_id=1 type=VOTER_FULL state=StateReplicate next=2
          store_id=2 replica_id=2 type=VOTER_FULL state=StateReplicate next=2
          store_id=3 replica_id=3 type=VOTER_FULL state=StateSnapshot next=2
          store_id=4 replica_id=4 type=VOTER_FULL state=StateReplicate next=1
          store_id=5 replica_id=5 type=VOTER_FULL state=StateReplicate next=1
        ----
        r1: [(n1,s1):1*,(n2,s2):2,(n3,s3):3,(n4,s4):4,(n5,s5):5]
    datadriven.go:144: 
        /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/rac2/rac2_test_/rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/force_flush_choice:103:
        stream_state [1 args]
        <no input to command>
        ----
        (n1,s1):1: state=replicate closed=false inflight=[1,2) send_queue=[2,2) precise_q_size=+0 B
        eval deducted: reg=+1.0 MiB ela=+0 B
        eval original in send-q: reg=+0 B ela=+0 B
        NormalPri:
          term=1 index=1  tokens=1048576
        ++++
        (n2,s2):2: state=replicate closed=false inflight=[1,2) send_queue=[2,2) precise_q_size=+0 B
        eval deducted: reg=+1.0 MiB ela=+0 B
        eval original in send-q: reg=+0 B ela=+0 B
        NormalPri:
          term=1 index=1  tokens=1048576
        ++++
        (n3,s3):3: closed
        ++++
        (n4,s4):4: state=replicate closed=false inflight=[1,1) send_queue=[1,2) precise_q_size=+1.0 MiB watching-for-tokens
        eval deducted: reg=+0 B ela=+1.0 MiB
        eval original in send-q: reg=+1.0 MiB ela=+0 B
        ++++
        (n5,s5):5: state=replicate closed=false inflight=[1,1) send_queue=[1,2) precise_q_size=+1.0 MiB force-flushing
        eval deducted: reg=+0 B ela=+1.0 MiB
        eval original in send-q: reg=+1.0 MiB ela=+0 B
        ++++
        schedule-controller-event-count: 1
        scheduled-replicas: 5
    datadriven.go:333: 
        panic during /var/lib/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/bin/pkg/kv/kvserver/kvflowcontrol/rac2/rac2_test_/rac2_test.runfiles/com_github_cockroachdb_cockroach/pkg/kv/kvserver/kvflowcontrol/rac2/testdata/range_controller/force_flush_choice:130:
        
    --- FAIL: TestRangeController/force_flush_choice (0.14s)
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRangeController.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go
+++ b/pkg/kv/kvserver/kvflowcontrol/rac2/range_controller_test.go
@@ -1322,7 +1322,11 @@ func TestRangeController(t *testing.T) {
 					var sizeCount, sizeBytes int64
 					for _, rcState := range state.ranges {
 						stats := RangeSendStreamStats{}
-						rcState.rc.updateSendQueueStatsRaftMuRCLocked(state.ts.Now())
+						func() {
+							rcState.rc.mu.Lock()
+							defer rcState.rc.mu.Unlock()
+							rcState.rc.updateSendQueueStatsRaftMuRCLocked(state.ts.Now())
+						}()
 						rcState.rc.SendStreamStats(&stats)
 						count, bytes := stats.SumSendQueues()
 						sizeCount += count
@@ -1366,7 +1370,11 @@ func TestRangeController(t *testing.T) {
 
 				r := state.ranges[roachpb.RangeID(rangeID)]
 				if refresh {
-					r.rc.updateSendQueueStatsRaftMuRCLocked(state.ts.Now())
+					func() {
+						r.rc.mu.Lock()
+						defer r.rc.mu.Unlock()
+						r.rc.updateSendQueueStatsRaftMuRCLocked(state.ts.Now())
+					}()
 				}
 				stats := RangeSendStreamStats{}
 				r.rc.SendStreamStats(&stats)

